web/modules/custom/dpl_url_proxy/src/Form/ProxyUrlConfigurationForm.php

```diff
@@ -0,0 +1,303 @@
+        '#limit_validation_errors' => [],
+        '#ajax' => [
+          'callback' => '::addmoreCallback',
+          'wrapper' => 'hostnames-fieldset-wrapper',
+        ],
+      ];
+    }
+
+    $form['hostnames']['actions'] = [
+      '#type' => 'actions',
+    ];
+    $form['hostnames']['actions']['add'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Add one more', [], $t_opts),
+      '#submit' => ['::addOne'],
+      '#limit_validation_errors' => [],
+      '#ajax' => [
+        'callback' => '::addmoreCallback',
+        'wrapper' => 'hostnames-fieldset-wrapper',
+      ],
+    ];
+
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Submit', [], $t_opts),
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId(): string {
+    return 'form_api_example_ajax_addmore';
+  }
+
+  /**
+   * Callback for the "Add one more" button.
+   *
+   * @param mixed[] $form
+   *   Drupal form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Drupal form state object.
+   *
+   * @return mixed[]
+   *   THe "hostnames" fieldset.
+   */
+  public function addmoreCallback(array &$form, FormStateInterface $form_state): array {
+    return $form['hostnames'];
+  }
+
+  /**
+   * Adds one more host name element.
+   *
+   * @param mixed[] $form
+   *   Drupal form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Drupal form state.
+   */
+  public function addOne(array &$form, FormStateInterface $form_state): void {
+    $indexes = $form_state->get('indexes');
+    $last = end($indexes);
+    $indexes[] = $last + 1;
+    $form_state->set('indexes', $indexes);
+    $form_state->setRebuild();
+  }
+
+  /**
+   * Removed one host name element.
+   *
+   * @param mixed[] $form
+   *   Drupal form array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   Drupal form state.
+   */
+  public function removeOne(array &$form, FormStateInterface $form_state): void {
+    if ($triggerring_element = $form_state->getTriggeringElement()) {
+      $remove_value = $triggerring_element['#name'];
+      $key = array_search($remove_value, $form_state->get('indexes'));
+      if ($key !== FALSE) {
+        unset($form_state->get('indexes')[$key]);
+      }
+      $form_state->setRebuild();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state): void {
+    $values = [];
+    if ($form_state->getValue('prefix')) {
+      $values['prefix'] = $form_state->getValue('prefix');
+    }
+    $values['hostnames'] = array_reduce($form_state->getValue(['hostnames']),
+      function ($carry, $item) {
+        if (!empty($item['hostname'])) {
+          unset($item['remove_this']);
+          $carry[] = $item;
+        }
+        return $carry;
+      }, []);
+
+    $this->config('dpl_url_proxy.settings')
+      ->set('values', $values)
```

Comment:
I considered if we could do better than simply storing arrays of values. With [PHP 8.0 and constructor property promotion](https://stitcher.io/blog/constructor-promotion-in-php-8) (and read-only properties in PHP 8.1) we have a pretty simple way to create value objects. However then we would need some way to serialize/deserialize these objects from the configuration. Then I thought out this would do for now :smile:.

Reply:
👍